### ISSUE TITLE
Multicast Support

### ISSUE THREAD

**Issue by bmah888** [esnet/iperf#27]
_From [kennethwebber](https://code.google.com/u/105865501134735294197/) on August 11, 2010 14:45:48_

Explaination of new feature Support for Multicast testing. Justification of new feature A lot of video is multicast and has QOS implication that would desire testing.

_Original issue: http://code.google.com/p/iperf/issues/detail?id=27_

**Comment by wifiprintguy** [comment]
+1 for this enhancement

**Comment by elsticky** [comment]
+1 from me too! :)

**Comment by wifiprintguy** [comment]
I'm troubled by this issue being labeled as an "enhancement" - this is a feature in iperf2 that is lacking in iperf3, so it seems more appropriate to me to label this a defect than an enhancement.

**Comment by wifiprintguy** [comment]
IPv4 and IPv6 should both be supported whenever multicast support is added.  (Issue #442 requests IPv6 multicast support specifically but that is really a duplicate of this issue.

**Comment by bms** [comment]
Suggestions: RTP+RTCP support would be really useful. I have noticed MPEG4 TS streams don't normally ship with RTCP, so there's no way to prioritize the time base packets -- they are part of the same payload.

**Comment by nicorusti** [comment]
Any update about multicast support? +1 from me as well

**Comment by bmah888** [comment]
Just to be clear, nobody is working on this right now.  Contributions of code would be appreciated, but I'd urge anyone who's thinking about it to post their design before actually writing code, it's probably not going to be as easy as you think, primarily because the design of iperf3 is to have one client talking to one server (i.e. one-to-one communication).  That was a design choice.

**Comment by wifiprintguy** [comment]
@bmah888, can you summarize the challenges that a contributor would be facing to adding multicast support? When you say "the design of iperf3 is to have one client talking to one server", why does multicast support cause a problem? 

Multicast support is very important for profiling network problems in Wi-Fi environments, where the multicast traffic is delivered via an unreliable (non-acked) frame address type. Many discovery protocols above layer 3 use multicast extensively.

**Comment by harbinger0x7c0** [comment]
+1 for this enhancement
+1 for wifiprintguy's comment that this should be a defect rather than an enhancement
+1 for wifiprintguy's question regarding the challenges for adding multicast support

After going to the effort of bringing iperf into an isolated network for multicast testing, followed by no small amount of consternation over failing to make it behave as I expected, I was quite disappointed to learn that iperf3 not only doesn't support multicast, but also couldn't be bothered to tell me as much.

**Comment by ZJamesHan** [comment]
so what's the alternative tool to test multicast in a high-bandwidth and low-latency  environment ?  iperf seems to limit to under 900 Mbps

**Comment by bmah888** [comment]
@ZJamesHan : iperf2 has the multicast functionality you want, and is once again under active development.

**Comment by beau-williamson** [comment]
@bmah888 
Bruce,

If there is no plans to fix the "defect" (yes it **is** a defect IMHO) of no Multicast in **iperf3** and the answer is to use **iperf2** which is "_once again under active development_", then how about updating the iperf Web Site to show that **iperf2** is NOT Deprecated and is a supported package. (Otherwise you can't really say that it is _once again under active development_.)

Forgive me for saying this but if the new structure of **iperf3** was based on a "design choice" that resulted in making it nearly impossible to add Multicast support, then that was a really bad idea.  You've seen from comments in this thread that Multicast (both ASM and SSM) Testing was an important feature of **iperf2** for many networks and testers. (Someone correct me if I'm wrong but it was pretty much the only OpenSource option for testing Multicast.) With the addition of Noushi's SSM enhancement **iperf2** became pretty much the only _**SSM Multicast**_ test option in town. (Again, at least to my knowledge.) Once more, at the very least can someone update the iperf Web Site to indicate that support for **iperf2** is still there, active development is on-going and if it hasn't already been merged into **iperf2**, could someone merge the SSM enhancement into the latest **iperf2** code.

Respectfully,

Beau Williamson
aka Mr.Multicast

**Comment by nicorusti** [comment]
From my side I confirm multicast is needed for testing, for example, IPTV flows.

**Comment by bmah888** [comment]
@beau-williamson :  To what "iperf Website" are you referring?  I can think of several, but the only places where I can see where iperf2 is mentioned as being deprecated or unmaintained are:

http://iperf.sourceforge.net/
https://sourceforge.net/projects/iperf/

I will try to update those (there may be some issues of access) to reflect that a third party has taken over development of iperf2.

http://software.es.net/iperf/ is really the main iperf3 Web site (it's the one that gets updated regularly).  I don't see anyplace there in which we describe iperf2 as being unmaintained, in fact we specifically give a pointer to the fork / take-over of iperf2 in the FAQ:

http://software.es.net/iperf/faq.html

As for other sites, note that http://iperf.fr/ seems to be an enthusiast site not maintained by any of the iperf2 or iperf3 developers (in fact as of when I last checked it wasn't being updated at all and its maintainer was not answering email).  For some reason it shows up at the top of my search results for iperf3 using various search engines.  I apologize if that causes confusion, but I'm not sure what is the best way to handle this.

Note that iperf2 seems to be supported by someone who's not one of the original iperf2 or iperf3 developers, so you can hardly blame the iperf3 maintainer for something that someone in the community decided to pick up.  He's actually doing a fine job, but the development of iperf2 and iperf3 are proceeding largely in parallel.  We don't coordinate much on these projects.

Please note that iperf3 was developed for specific uses within ESnet and the perfSONAR / R&E networking community.  We make it available to the larger community as well, even those who have very different needs than us, and on a best effort basis we try to accommodate those other's requests.  Occasionally there are some features (such as multicast) that are clearly outside our scope and we just can't devote our limited resources to working on it.  If you don't like the lack of multicast functionality of iperf3, that's fine...feel free to use iperf2 or any of the other bandwidth testing tools available to you, or fork the iperf3 code and implement it yourself.  There are a number of alternatives and I encourage you to find the one that works best for you.

**Comment by beau-williamson** [comment]
Bruce,

 

Yes, there is an abundance of confusion on this due to a variety of reasons. On reason is that iperf3 is under GitHub titled simply iperf and not iperf3 and there doesn’t appear to be anything to point people over to iperf2 for those still needing it. People that need the functionality of iperf2 and that are trying to find it by looking for iperf2 on GitHub won’t find anything.  I think it might be helpful to add some text in the GitHub iperf README.md file (maybe near the “Changes from iperf2.x” section) that states that iperf2 is deprecated but is still being maintained on Source Forge by others that depend on key features in iperf2.  Then add a link back over to Source Forge so folks needing those features can find the latest iperf2 code.

 

It would be nice if the iperf2 supporters on Source Forge would migrate things to GitHub under the title of iperf2.  That might also make it easier to merge the Noushi SSM Multicast enhancement that is published in GitHub with the on-going work on iperf2.  I suspect however, that that would be a bit much to ask to make that migration.  But who knows.  (It never hurts to ask.)

 

Also, if the folks at http://software.es.net/iperf/ could add some text on the main page (not just the FAQ) that mentions iperf2 and adds a link to the https://sourceforge.net/projects/iperf/ page for those wanting iperf2 and its features, that too would be very helpful.

 

And finally, if at all possible, if someone could update the http://iperf.sourceforge.net/ page to say that iperf2 can be found at https://sourceforge.net/projects/iperf/ for those still needing some of the features of iperf2 that were dropped in iperf3, that too would help cut down on some of the iperf2/iperf3 confusion.  

 

Thanks for your help.

 

Beau Williamson

**Comment by bmah888** [comment]
@beau-williamson : Thanks for the suggestions!  I got access to the iperf repository on SourceForge, and put pointers in a couple of places to iperf2 repository on Sourceforge (also fixed an obsolete link to Google Code).  I agree it'd probably make sense to add those same links in the README.md file for iperf3, so that's now on my TODO list.

The only thing I can't help with is the moving of iperf2 to GitHub.  That's something that'll have to come from the maintainer of iperf2.  I am not sure what his view is on this...I'm guessing he kept things on SF because that's where we had the original iperf repository.

We probably could have done a better job naming the iperf3 repositories (like actually calling them "iperf3").  I think the naming came from a desire to just in essence make iperf2 just go away, but admittedly that's hard to do with an open source project whose successor doesn't have complete feature parity, and our expectations in that area were probably not completely realistic.  I'm glad that someone took up maintenance of iperf2 because we've learned there really is a community for that work.

**Comment by beau-williamson** [comment]
Bruce,

 

Thanks for helping reduce the confusion of iperf2 and iperf3 by doing those updates.  Once you get the README.md file updated that will really help.

 

As for getting the iperf2 guys to migrate over to GitHub, I seriously doubt that will happen.  They already have a substantial Ticket DB and other SourceForge infrastructure setup for iperf2 so it doesn’t make much sense to try to tackle the huge job of migrating.  Better to devote their volunteer cycles to working on iperf2 code.

 

Maybe someday down the road someone on the iperf3 team will add back the IP Multicast stuff AND include the extension for SSM Multicast that is so badly needed.  I understand that may be difficult given the new architecture of your code but I can always hope.  😊

 

Best Regards,

 

Beau Williamson

aka Mr. Multicast

**Comment by rjmcmahon** [comment]
Sorry, just saw this thread.  I'll look at adding SSM to iperf2 soon.  Moving from sourceforge to github is lower priority then feature updates (though I hear the request.)

Bob

**Comment by bmah888** [comment]
Closing.  This is a feature that we're choosing not to support.  Users who want to do throughput testing on multicast groups should use iperf2 (or some other suitable tool).
